### DISCUSSION TITLE
I created Bitmap Graphics Library for RP6502

### DISCUSSION THREAD

**Discussion by vruumllc** [picocomputer/community#53] [Q&A]
![bitmap_graphics](https://github.com/picocomputer/discussions/assets/145504135/e515e86f-494c-4d36-a5ed-e2a146088238)
Here is a simple bitmap graphics library and demo program that I created for the RP6502:

https://github.com/vruumllc/RP6502-Bitmap_Graphics

It makes creating simple background graphics easy for all the bitmap modes (mode3) the RP6502 supports.
Updated for new CC65 compiler with rp6502 support.
Enjoy!

**Comment by rumbledethumps** [comment]
If you want to eek out that last bit of performance before jumping into assembly, check this: https://github.com/ilmenit/CC65-Advanced-Optimizations

**Reply by vruumllc** [reply]
Thanks, interesting reading. It will be fun to figure out how to make things as fast as possible. At first, I just focused on figuring out how to draw in all the different modes, and presenting it in as clear as possible way for other's to learn from. Thus the key-event code as well.  

The only thing that still stumps me is the color specification for 16-bits/pixel. The macros in the docs didn't seem to work, and so I did it by trial and error using a binary constant. It seemed to be 3-bits B,3-bits R, 3-bits G. Is that right? Could you fill this properly for me: 
MSB> ???? ???B BBRR RGGG <LSB. I know there is supposed to be an alpha (A) bit.

**Reply by rumbledethumps** [reply]
BGAR5515 - BBBBBGGGGGARRRRR

**Reply by vruumllc** [reply]
Yeah, that is what the macro did. It didn't work for me. I'll have to go back and look at it again. Thanks.

**Reply by rumbledethumps** [reply]
Here's your error.
```
    if (bpp == 16) {
        RIA.addr0 = (canvas_width() * y + x) * 2;
        RIA.step0 = 1;
        RIA.rw0 = color;
        RIA.rw0 = color;
```

**Reply by vruumllc** [reply]
I checked it out, but don't see anything wrong with that code. 
I stripped down your mode3 demo to just draw 16bpp, and the macros still don't work.
Could you run it for me? Thanks!
------------------------------------------------------------------------------------------------------------------
/*
 * Copyright (c) 2023 Rumbledethumps
 *
 * SPDX-License-Identifier: BSD-3-Clause
 * SPDX-License-Identifier: Unlicense
 */

#include <rp6502.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <stdint.h>
#include <fcntl.h>

#define COLOR_FROM_RGB5(r,g,b) ((b<<11)|(g<<6)|(r))
#define COLOR_ALPHA_MASK (1u<<5)

// This is not good code to learn from,
// it's a functional test for developing mode3.

void clear()
{
    unsigned i;
    RIA.addr0 = 0;
    RIA.step0 = 1;
    for (i = 0; i < 61440u / 8; i++)
    {
        RIA.rw0 = 0;
        RIA.rw0 = 0;
        RIA.rw0 = 0;
        RIA.rw0 = 0;
        RIA.rw0 = 0;
        RIA.rw0 = 0;
        RIA.rw0 = 0;
        RIA.rw0 = 0;
    }
}

void box(int color, uint8_t bpp)
{
    int x1, y1, x2, y2, x, y;
    x1 = ((uint32_t)rand() * 240) >> 15;
    y1 = ((uint32_t)rand() * 128) >> 15;
    x2 = ((uint32_t)rand() * 240) >> 15;
    y2 = ((uint32_t)rand() * 128) >> 15;
    if (x1 > x2)
    {
        x = x1;
        x1 = x2;
        x2 = x;
    }
    if (y1 > y2)
    {
        y = y1;
        y1 = y2;
        y2 = y;
    }

    // bpp = 16
    RIA.step0 = 1;
    for (y = y1; y < y2; y++)
    {
        RIA.addr0 = 480 * y + x1; // I think X1 should be multiplied by 2 here
        for (x = x1; x < x2; x++)
        {
            RIA.rw0 = color;
            RIA.rw0 = color;
        }
    }

    // wait for keypress
    RIA.addr0 = 0xFF10;
    RIA.step0 = 0;
    while (RIA.rw0 & 1)
        ;
    return;
}

void main()
{
    srand(6);
    _randomize();

    xreg_vga_canvas(2);
    clear();

    xram0_struct_set(0xFF00, vga_mode3_config_t, x_wrap, false);//true);
    xram0_struct_set(0xFF00, vga_mode3_config_t, y_wrap, false);//true);
    xram0_struct_set(0xFF00, vga_mode3_config_t, x_pos_px, 0);
    xram0_struct_set(0xFF00, vga_mode3_config_t, y_pos_px, 0);
    xram0_struct_set(0xFF00, vga_mode3_config_t, width_px, 240);
    xram0_struct_set(0xFF00, vga_mode3_config_t, height_px, 128);
    xram0_struct_set(0xFF00, vga_mode3_config_t, xram_data_ptr, 0x0000);
    xram0_struct_set(0xFF00, vga_mode3_config_t, xram_palette_ptr, 0xFFFF);

    xreg_vga_mode(0, 1, 172, 180);
    xreg_ria_keyboard(0xFF10);
    xreg_vga_mode(3, 4, 0xFF00, 0, 0, 172);

    printf("\nBGAR5515 Macro RED box.(press any key)");
    clear();
    box(COLOR_FROM_RGB5(31,0,0)|COLOR_ALPHA_MASK, 16);

    printf("\nBGAR5515 Macro GRN box.(press any key)");
    clear();
    box(COLOR_FROM_RGB5(0,31,0)|COLOR_ALPHA_MASK, 16);

    printf("\nBGAR5515 Macro BLU box.(press any key)");
    clear();
    box(COLOR_FROM_RGB5(0,0,31)|COLOR_ALPHA_MASK, 16);

    printf("\nBGAR5515 Binary RED box.(press any key)");
    clear();
    box(0b0000000000111111, 16);

    printf("\nBGAR5515 Binary GRN box.(press any key)");
    clear();
    box(0b0000011111100000, 16);

    printf("\nBGAR5515 Binary BLU box.(press any key)");
    clear();
    box(0b1111100000100000, 16);

    printf("\nBRG333 Binary RED box.(press any key)");
    clear();
    box(0b0000000000111000, 16);

    printf("\nBRG333 Binary GRN box.(press any key)");
    clear();
    box(0b0000000000000111, 16);

    printf("\nBRG333 Binary BLU box.(press any key)");
    clear();
    box(0b0000000111000000, 16);
    printf("\n");
}

**Reply by vruumllc** [reply]
Sorry about the weird BOLD rendering above. I didn't mean to scream. 8^(

**Reply by rumbledethumps** [reply]
Try this:
```
    if (bpp == 16) {
        RIA.addr0 = (canvas_width() * y + x) * 2;
        RIA.step0 = 1;
        RIA.rw0 = color;
        RIA.rw0 = color >> 8; // You forgot to shift
```

**Reply by vruumllc** [reply]
Excellent! That did the trick.
It looks like the bugs are in your mode3 demo code, too, which I just copied.

**Reply by rumbledethumps** [reply]
Probably lots of other things in that file that turn into bugs after copypasta.
```
// This is not good code to learn from,
// it's a functional test for developing mode3.
```
The documentation is the only source of truth.

**Reply by vruumllc** [reply]
Indeed. Unfortunately, I'm a (lazy?) learn-by-example kind of guy.

BTW, I saw the >> 8 fix on github, but I still think you need to multiply x1*2 as well:
RIA.addr0 = 480 * y + x1; // I think X1 should be multiplied by 2 here

Thanks so much for your patience and help! I am having so much fun with this project. I'm working on some optimizations now.

**Reply by vruumllc** [reply]
I have replaced the bitmap_graphics.zip file, above, with a corrected version.

**Comment by vruumllc** [comment]
I have added my Bitmap Graphics library to [my GitHub repositories](https://github.com/vruumllc), and have modified the initial post to point to it. 

I have also added a version of the library for the LLVM-MOS C compiler.
